ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/TraceRawTransaction.java

```diff
@@ -0,0 +1,134 @@
+
+import java.util.Optional;
+import java.util.Set;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class TraceRawTransaction implements JsonRpcMethod {
+  private static final Logger LOG = LoggerFactory.getLogger(TraceRawTransaction.class);
+  private final TransactionSimulator transactionSimulator;
+  private final BlockchainQueries blockchainQueries;
+  private final ProtocolSchedule protocolSchedule;
+
+  public TraceRawTransaction(
+      final ProtocolSchedule protocolSchedule,
+      final BlockchainQueries blockchainQueries,
+      final TransactionSimulator transactionSimulator) {
+    this.transactionSimulator = transactionSimulator;
+    this.protocolSchedule = protocolSchedule;
+    this.blockchainQueries = blockchainQueries;
+  }
+
+  @Override
+  public String getName() {
+    return transactionSimulator != null ? RpcMethod.TRACE_RAW_TRANSACTION.getMethodName() : null;
+  }
+
+  @Override
+  public JsonRpcResponse response(final JsonRpcRequestContext requestContext) {
+    if (requestContext.getRequest().getParamLength() != 2) {
+      return new JsonRpcErrorResponse(
+          requestContext.getRequest().getId(), JsonRpcError.INVALID_PARAMS);
+    }
+
+    final String rawTransaction = requestContext.getRequiredParameter(0, String.class);
+
+    final Transaction transaction;
+    try {
+      transaction = DomainObjectDecodeUtils.decodeRawTransaction(rawTransaction);
+    } catch (final RLPException | IllegalArgumentException e) {
+      return new JsonRpcErrorResponse(
+          requestContext.getRequest().getId(), JsonRpcError.INVALID_PARAMS);
+    }
+
+    LOG.trace("Received local transaction {}", transaction);
+
+    // second param is the trace type/s
+    final TraceTypeParameter traceTypeParameter =
+        requestContext.getRequiredParameter(1, TraceTypeParameter.class);
+
+    final Set<TraceTypeParameter.TraceType> traceTypes = traceTypeParameter.getTraceTypes();
+    final DebugOperationTracer tracer = new DebugOperationTracer(buildTraceOptions(traceTypes));
+    final Optional<TransactionSimulatorResult> maybeSimulatorResult =
+        transactionSimulator.process(
+            CallParameter.fromTransaction(transaction),
+            buildTransactionValidationParams(),
+            tracer,
+            blockchainQueries.headBlockNumber());
+
+    if (maybeSimulatorResult.isEmpty()) {
+      throw new IllegalStateException("Invalid transaction simulator result.");
+    }
+
+    final TransactionTrace transactionTrace =
+        new TransactionTrace(
+            maybeSimulatorResult.get().getTransaction(),
+            maybeSimulatorResult.get().getResult(),
+            tracer.getTraceFrames());
+    final Block block = blockchainQueries.getBlockchain().getChainHeadBlock();
+
+    Object response =
+        TraceUtils.getTraceCallResult(
+            protocolSchedule, traceTypes, maybeSimulatorResult, transactionTrace, block);
+
+    if (response instanceof JsonRpcErrorResponse) {
```

Comment:
Can response be a JsonRpcErrorResponse in itself? Below we had to construct a JsonRpcSuccessResponse with response, typing the response as Object rather than JsonNode also doesn't feel quite right?

Reply:
I think you're right. copied that code out of context. fixed.